integrations/anthropic/src/actions/generate-content.ts

```diff
@@ -0,0 +1,261 @@
+      temperature: input.temperature,
+      top_p: input.topP,
+      system: input.systemPrompt,
+      stop_sequences: input.stopSequences,
+      metadata: {
+        user_id: input.userId,
+      },
+      tools: mapToAnthropicTools(input),
+      tool_choice: mapToAnthropicToolChoice(input.toolChoice),
+      messages,
+    })
+  } catch (err) {
+    logger.forBot().error(`generateContent action failed because Anthropic returned an error: ${JSON.stringify(err)}`)
+    throw err
+  }
+
+  const { input_tokens: inputTokens, output_tokens: outputTokens } = response.usage
+
+  const content = response.content
+    .filter((x): x is Anthropic.TextBlock => x.type === 'text') // Claude models only return "text" or "tool_use" blocks at the moment.
+    .map((content) => content.text)
+    .join('\n\n')
+
+  return <llm.schemas.GenerateContentOutput>{
+    id: response.id,
+    provider: 'anthropic',
+    model: response.model,
+    choices: [
+      // Claude models don't support multiple "choices" or completions, they only return one.
+      {
+        role: response.role,
+        type: 'multipart',
+        index: 0,
+        stopReason: mapToStopReason(response.stop_reason),
+        toolCalls: mapToToolCalls(response),
+        content,
+      },
+    ],
+    usage: {
+      inputTokens,
+      inputCost: calculateTokenCost(modelSpecs.inputCostPer1MTokens, inputTokens),
+      outputTokens,
+      outputCost: calculateTokenCost(modelSpecs.outputCostPer1MTokens, outputTokens),
+    },
+  }
+}
+
+function calculateTokenCost(costPer1MTokens: number, tokenCount: number) {
+  return (costPer1MTokens / 1_000_000) * tokenCount
+}
+
+async function mapToAnthropicMessage(message: llm.schemas.Message): Promise<Anthropic.MessageParam> {
+  return {
+    role: message.role,
+    content: await mapToAnthropicMessageContent(message),
+  }
+}
+
+async function mapToAnthropicMessageContent(message: llm.schemas.Message): Promise<Anthropic.MessageParam['content']> {
+  if (message.type === 'text') {
+    if (typeof message.content !== 'string') {
+      throw new InvalidPayloadError('`content` must be a string when message type is "text"')
+    }
+
+    return message.content as string
+  } else if (message.type === 'multipart') {
+    if (!Array.isArray(message.content)) {
+      throw new InvalidPayloadError('`content` must be an array when message type is "multipart"')
+    }
+
+    const content: Anthropic.MessageParam['content'] = []
+
+    for (const part of message.content) {
+      if (part.type === 'text') {
+        content.push(<Anthropic.TextBlockParam>{
+          type: 'text',
+          text: part.text,
+        })
+      } else if (part.type === 'image') {
+        if (!part.url) {
+          throw new InvalidPayloadError('`url` is required when part type is "image"')
+        }
+
+        if (!part.mimeType) {
+          throw new InvalidPayloadError('`mimeType` is required when part type is "image"')
+        }
```

Comment:
:question:**question:** These validations seems duplicated in llm/openai, shouldn't we just have one validation done with the definition of the generateContent validation, since it's the same for all llms?

Reply:
Yes, some some of these validations are being duplicated. My plan was that once a 3rd official LLM integration is done (e.g. Google AI, for the Gemini models) I'll introduce a pre-validator that can be reused across these 3 integrations. I'd also wait to introduce this until tests are written for these LLM integrations to make sure nothing breaks.

Reply:
BTW, about this specific validation for the presence of mimeType, it's only applicable so far to Anthropic because OpenAI allows just passing an actual URL to the image and they seem to extract the mimeType from the HTTP response to that URL.

Reply:
I made the MIME type optional as I realized we can simply detect it when fetching the image, here's the commit: https://github.com/botpress/botpress/pull/13159/commits/1dd9056a7f67a6d8fd3c80ebf8a46e637deae734